gef.py

```diff
@@ -6760,55 +6749,111 @@ def do_invoke(self, argv):
             info("No Tcache in this version of libc")
             return
 
-        arena = GlibcArena("*{:s}".format(argv[0])) if len(argv) == 1 else get_main_arena()
-
-        if arena is None:
-            err("Invalid Glibc arena")
+        current_thread = gdb.selected_thread()
+        if current_thread is None:
+            err("Couldn't find current thread")
             return
 
-        # Get tcache_perthread_struct for this arena
-        heap_base = HeapBaseFunction.heap_base()
-        if heap_base is None:
-            err("No heap section")
-            return
-        addr = heap_base + 0x10
-
-        gef_print(titlify("Tcachebins for arena {:#x}".format(int(arena))))
-        for i in range(GlibcArena.TCACHE_MAX_BINS):
-            if get_libc_version() < (2, 30):
-                count = ord(read_memory(addr + i, 1))
+        threads = gdb.selected_inferior().threads()
+        if argv:
+            if "all" in argv:
+                tids = [t.num for t in threads]
             else:
-                count = u16(read_memory(addr + 2 * i, 2))
-            chunk = arena.tcachebin(i)
-            chunks = set()
-            m = []
+                tids = self.check_thread_ids(argv)
+        else:
+            tids = [current_thread.num]
 
-            # Only print the entry if there are valid chunks. Don't trust count
-            while True:
-                if chunk is None:
-                    break
+        # As a nicety, we want to display threads in ascending order by gdb number
+        for thread in sorted(threads, key=lambda t: t.num):
+            if thread.num not in tids:
+                continue
 
-                try:
-                    m.append("{:s} {:s} ".format(LEFT_ARROW, str(chunk)))
-                    if chunk.address in chunks:
-                        m.append("{:s} [loop detected]".format(RIGHT_ARROW))
+            thread.switch()
+            tcache_addr = gdb.parse_and_eval("(void *) tcache")
+
+            gef_print(titlify("Tcachebins for thread {:d}".format(thread.num)))
+            for i in range(self.TCACHE_MAX_BINS):
+                chunk, count = self.tcachebin(int(tcache_addr), i)
```

Comment:
If the breakpoint occurs before the `tcache_perthread_struct` is initialized, `tcache_addr` will be 0, giving an ambiguous error saying:

```
[!] Command 'heap bins tcache' failed to execute properly, reason: can't convert negative int to unsigned
```

Try the following test case:

```
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void *thread1(void *vargp)
{
    char* a = (char*) malloc(40);
    char* b = (char*) malloc(40);
    char* c = (char*) malloc(40);
    free(a);
    free(b);
    free(c);
    return NULL;
}

int main()
{
    pthread_t thread_id1;
    pthread_create(&thread_id1, NULL, thread1, NULL);
    __builtin_trap();
    exit(0);
}
```

This is not so likely to occur, but it would be good to handle this case by checking if `tcache_addr == 0` before proceeding.

Reply:
I couldn't actually reproduce the issue, but the problem makes sense conceptually. Should be fixed in a6852ff1c5e7f8c9a5175184844bb6f9c3a0db62